[SIL Optimizer] Replace VarDeclUsageChecker AST Walker with DiagnoseVarUsage SIL Pass

maustinstar opened the pull request:
# SIL Diagnostics for Var Usage

This mandatory SIL optimization pass, `DiagnoseVarUsage`, replaces `VarDeclUsageChecker` as the system for generating usage diagnostics. 

#### Why a SIL pass?
With access to the def-use graph, we have a more accurate picture of the lifetime of a variable declaration for more robust diagnostics. Ultimately, the former AST walker missed many cases, shown in the bugs below. Additionally, this implementation has many possibilities for future features.

## Addresses Bugs:

### Mutating Inouts

```swift
func increment() {
  var x = 0 // no warning, but should warn x was mutated but never read.
  x += 1
}
```
This bug (and similar bugs) are addressed by improving the way usage of inout parameters is diagnosed. If the function is serialized or local to the SIL module, we can follow the SILArgument to look for a read of the parameter. For external SILFunctions, an inout parameter is marked as used if the function produces some used output, or if the function has a closure parameter (where the inout argument _may_ be used to produce other parameters within the closure).

### Closures Missing Diagnostics

SR-11992, SR-13821

```swift
class Foo {
  lazy var x: Int = {
    var y = 42 // no warning, but should warn unused
    var z = 42 // no warning, but should warn never mutated
    return z
  }()
}
```

```swift
var missesAllDiagsInClosures: () -> Int = {
  var unused = 42 // no warning, but should warn unused
  return 12
}
```

Bugs like this are addressed because we are no longer using the AST walker, which inconsistently missed certain nodes. Instead, we can see every SILInstruction that originates from a VarDecl and then decide whether to inspect it for diagnostics.

### Misleading Unused Setter Param Diagnostics

```swift
var x : Int {
  get { 0 }
  set { }
}
x
// warning: setter argument 'newValue' was never used, but the property was accessed
// note: did you mean to use 'newValue' instead of accessing the property's current value?
```

These diagnostics are misleading because `newValue` is not in scope for the diagnostic location, and the setter function was never called. Now, a case like this does not produce diagnostics.

Cases like the following still produce diagnostics as expected.

```swift
var x : Int {
  get { 0 }
  set {
    print(x)
    // warning: setter argument 'newValue' was never used, but the property was accessed
    // note: did you mean to use 'newValue' instead of accessing the property's current value?
  }
}
```

> For some more information on this decision, [this gist](https://gist.github.com/maustinstar/7cfcbe12b96ee4317bab2445931612a3) details reasoning and a possible alternative diagnostic.

***

And many other similar cases are addressed...

After switching to the SIL implementation, we now get consistent diagnostics in many of these common cases.

## Modifications and Limitations

### Test files

This diagnostic occurs during the mandatory diagnostics SIL optimizations, and therefore the diagnostics are only emitted after errors are addressed in the file (because code with errors does not reach SILGen).

Nearly every test case regarding variable usage is touched by this PR. I apologize for the nightmare this review may entail.

If I removed expected warnings, it is because the file never reaches SIL optimizations. Rest assured, every case I edited is either now checked in `SILOptimizer/diagnose_var_usage` or moved from the original `<testname>.swift` to `<testname>_usage.swift`.

### Discarded Assignments

We now track discarded assignments by removing the SILGen peephole for loading discarded values, or we use a new SILInstruction `mark_discarded` to note the usage of a trivial value like a discarded `DeclRefExpr`. This instruction is not canonical SIL, and has no semantic changes to SIL.

### #if declarations

Because inactive clauses are not visible in SIL, an AST walker exempts declarations used in inactive if-config clauses from usage diagnostics. The unused leading Bit of `TailAllocatedDebugVariable` was repurposed to track this diagnostic exemption.

```swift
var x = 12
#if false
  x = 1
#endif
print(x)
```

### Tracking Value Usage

In this SIL pass, we track the use of a *value*, not the *declaration*. This means if we have multiple constants referencing the same value, we have no way of knowing which reference to the value we used:

```swift
func duplicateReference() {
  let x = 12
  let y = x
  print(x) // we don't know if x or y was used here, because they are identical in SIL
}
```

To a small extent, we can check up through the AST to find if a user of the value corresponds to a certain reference, and we do this to at least know x was assigned to y.

To account for this situation, instead of generating possibly misleading usage diagnostics, we introduce a new diagnostic:

```swift
func duplicateReference() {
  let x = 12 // note: initially referenced here
  let y = x // warning: immutable value referenced by 'y' is already referenced by 'x'; consider removing it and replacing all uses of 'y' with 'x'}}
}
```

> Future work may be done here to improve this diagnostic to include fix-its for replacement. Or, if this diagnostic is to be reverted, we would need to produce separate values for every variable declaration.
> 
> What currently looks like this:
> ```
> debug_value %3 : $Int, let, name "x"            // id: %4
> debug_value %3 : $Int, let, name "y"            // id: %5
> ```
> 
> Would need to do something like this:
> ```
> debug_value %3 : $Int, let, name "x"
> ...
> copy_value %3 to %4
> debug_value %4 : $Int, let, name "y"
> ```

However, in the handful of times this pattern appears in the stlib, the suggestion of the diagnostic made more sense, was quick to address, and lead to more straight-forward code. 

### Tracking Value Usage - Exceptions

**Capture Lists** - redeclare an identical value, so we don't want to produce incorrect diagnostics here. We perform an explicit check for declarations belonging to capture lists to make sure this declaration of the value is captured.

**Case Statements** - binding declarations in switch statements may redeclare an identical value. Again, we explicitly check that a 'use' of the value is in SILBlocks accessible to the case statement.

## Future Extensions

While the focus of this PR is to lay the groundwork to replace the previous AST Walker implementation, I noticed this SIL pass opens up plenty of further diagnostics.

### Unused Parameter diags

See SR-2819

This pass already performs a check for the `newValue` parameter of setter functions, and the logic could easily be transferred to diagnosing other unused parameters or unmutated inout parameters. If the parameter is not required (such as when overriding a parent function), then these diagnostics could be helpful to simplify code.

### Unused Local Functions

Using def-use traversals in a similar manner for functions inaccessible to outside modules can produce useful diagnostics.

## Links
SR-11992 and SR-13821
rdar://59636530
Superceeds PR #34617

Thanks to @CodaFi  and @augusto2112 for the brainstorming sessions and motivation!

maustinstar committed:
[SR-11992] Check for unused properties in closures

maustinstar committed:
[WIP] Implement VarDeclUsageChecker as SIL Optimization Pass

maustinstar committed:
[WIP] SIL Pass for VDUC

Changed from recursive structure to iterative.
Handles most cases correctly, with minor misdiagnoses
in the stdlib.

Future work:
  - prevent more stdlib misdiagnoses
  - needs extensive testing for certain cases
    that may be making assumptions; see FIXMEs

maustinstar committed:
[WIP] Add, edit, and remove* test cases

Future work:
  - *removed cases must be examined and re-added
  - many tests need some more variants

maustinstar committed:
Remove reference to missing header

maustinstar committed:
[WIP] Generating test cases

maustinstar committed:
[WIP] Differentiate multiple references of same value

Multiple VarDecls may reference the same value; ensure the use of
the value matched the VarDecl identifier by examining the AST loc.

Functionality added for variable assignments.
Some test cases added.

Future work:
  - Check for variable Identifier inside ApplyInst (CallExpr)
  - more test cases

maustinstar committed:
Add diagnostic for multiple references to identical immutable value

maustinstar committed:
[WIP] Remove VDUC AST walker and refine VDU SIL Pass

Removes VDUC from misc diagnostics.
Implements extensinve support for captures and binding switches.
Adds test cases

maustinstar committed:
Create VarUsageInfo class for collecting diagnostic information.

maustinstar committed:
Add and modify test cases for DiagnoseVarUsage

slavapestov commented:
This is really cool!

xwu commented:
```swift
func increment() {
  var x = 0 // no warning, but should warn x was mutated but never read.
  x += 1
}
```

Well, _only_ if it’s a built-in `+=` do we know that. If I create a custom implementation that logs the previous value before incrementing by one, then the value _is_ both read and mutated. That is, admittedly, far-fetched for `+=`, but not for arbitrary inout functions.

maustinstar commented:
@xwu that's a great point. For custom inouts local to the SILModule, this code actually follows the argument to look for a read in the function body.

But your comment does make me think about custom functions that are declared externally, and you're right that we can't verify it was read.

I'd like to point out that if this inout function does not produce any used result, I'm not sure if the mutating function ever provided value to the use of the variable, and the warning *might* still make sense. But maybe this discussion is out of scope for this PR.

Perhaps, for now, I can change this to assume external inout functions read the parameter. How do you think we should approach this?

xwu commented:
> Constants have no getter function, so _ = x no longer counts as a use of the variable. This source code is invisible after SILGen.
>
>     let x = 1
>     _ = x
>     
>     /* SIL is identical to: */
>     
>     let x = 1
> 
> This is because the underscore assignments have no effect, and the value is never accessed.

While understandable, this does have some impact on the user experience, as one of the most obvious ways currently to silence the warning is to write `_ = x`, and it would be nice to preserve that for both mutable and immutable bindings.

xwu commented:
> Perhaps, for now, I can change this to assume external inout functions read the parameter. How do you think we should approach this?

Neat that you’re actually looking for a read in the function body where possible. I think it’s reasonable to scope the diagnostic so that it’s emitted only where the function body is available (either because it’s inlinable from the standard library or otherwise) or in the same module, and so you can be sure if there’s no read. That would cover a lot of scenarios, I think, without giving false positives.

maustinstar commented:
@xwu okay, thank you. I will make and test the appropriate changes for this.

I think it may be interesting to expand upon this in the future. Down the line, it would be great to provide these diagnostics for purely mutating functions, like the built-in `+=` operator.

xwu commented:
> Down the line, it would be great to provide these diagnostics for purely mutating functions, like the built-in `+=` operator.

Well, the built-in operator implementations are inlinable if not transparent, so you should be able to look through those and make the diagnosis, no?

maustinstar committed:
Inspect serialized SIL functions for inout var usage

External and non-serialized uses of inout parameters
will qualify as a read of the variable.

Internal or serialized uses will inspect the function
body for a read of the variable.

maustinstar commented:
@xwu thanks for the conversation! I made updates to inspect usage through transparent functions, and I think this is a much better approach to these diagnostics!

#### Summary of changes
- local or transparent functions are inspected for a read of inout parameters
- other external functions will assume a read of inout parameters

xwu commented:
> local or transparent functions

Is it too expensive to look through _any_ functions that have their function body serialized? I believe you can just use `isSerialized()` for that.

maustinstar commented:
Okay, I am running tests for that right now. This is my first time working with SIL, so I never knew that was possible. Thanks for the suggestion, and I'll tag you with an update after I test it!

maustinstar commented:
@xwu, this works! Thank you.

maustinstar commented:
Sorry I just noticed your comment about underscore assignments, @xwu. I realize this is not great behavior for these diagnostics. This change won't affect function results on the rhs, it only affects if an immutable value is directly used.

```swift
let x = 0
_ = x // this is a problem

_ = x + 1 // ok
_ = someFunctionResult() // ok
```

I will look into modifying SILGen to produce a `debug_value` instruction for underscore assignments to mark this as a read, but I have no idea what level of effort this entails. Will look into this tomorrow.

Thanks again!

xwu commented:
> I will look into modifying SILGen to produce a `debug_value` instruction for underscore assignments to mark this as a read, but I have no idea what level of effort this entails. Will look into this tomorrow.

Admittedly, I don't think it's a _huge_ deal since this is limited to constants; if it's that narrowly tailored, I don't know that many people are deliberately creating a binding and then not using it, but one thing I've learned is that if it's possible, someone is relying on it.

maustinstar committed:
Find discarded uses of constant values

Reverts a mis-diagnosis of unread let values
when assigned to an underscore

maustinstar commented:
@xwu, I made a fix to look for a `DiscardAssignmentExpr`, restoring the expected functionality of underscore assignments.

We do collect enough information at this stage to know the discarded assignment produces no effects, so it may be interesting to add a diagnostic here in the future.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@maustinstar Looking into callee function bodies sounds tricky because those functions might in turn call other functions, which might be recursive, etc. There is existing infrastructure in the SIL optimizer that could handle this if you decide to go down this route, but I wonder if its worth the added complexity. Also, it would be nice if we could avoid triggering deserialization of functions, since it can add a lot of unnecessary overhead at -Onone.

slavapestov commented on lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp:
```diff
@@ -1084,6 +1116,16 @@ class VariableReadTraverser {
+    
+    // As a last resort, check if the value was discarded in an
+    // underscore assignment.
+    if (auto AFD = dyn_cast<AbstractFunctionDecl>(getVarDecl(EntryInst)->
```
Walking the AST like this is a layering violation to me. At the very least, this won't handle closure bodies or other non-AbstractFunctionDecl local contexts.

slavapestov replied:
The reason `_ = foo` produces no code is because SILGen peepholes it away. Perhaps we should just not do that, and allow dead code elimination to remove the discarded result after the diagnostics pass?

maustinstar replied:
> Perhaps we should just not do that, and allow dead code elimination to remove the discarded result after the diagnostics pass

This definitely sounds like the ideal solution here. I will look into this. Thanks!

maustinstar replied:
This has been addressed in my latest commits! The peephole has been removed, and this makes uses of a `var` visible.. Even still, `let` constants do not have any instructions to generate, so I created a non-canonical `mark_discarded` instruction for cases like this.

maustinstar commented:
@slavapestov

> might in turn call other functions, which might be recursive

If your concern is that we might find ourselves in an infinite loop of searching through recursive functions, this won't happen because instructions that have already been seen are not searched again. 

Also, this def-use traversal will keep stepping into functions until a read is found. I don't know if performance is a huge concern here because this will only happen in the case that a variable gets used in multiple inout functions without ever reading the value.

> There is existing infrastructure in the SIL optimizer that could handle this if you decide to go down this route.

Great! Can you give me some pointers for where to look and learn?

> it would be nice if we could avoid triggering deserialization of functions

Can you explain a little further? In regards to inspecting callees, we only search serialized and locally defined SILFunctions.

Thank you for your feedback!

xwu commented:
> To handle diagnostics for this situation, instead of generating possibly misleading usage diagnostics, we introduce a new diagnostic:
> ```
> func duplicateReference() {
>   let x = 12 // note: initially referenced here
>   let y = x // warning: immutable value referenced by 'y' is already referenced by 'x'; consider removing it and replacing all uses of 'y' with 'x'}}
> }
> ```

I like that you detect this; it seems very much in keeping with detecting unused bindings. One caveat and one nit:

1. There is a reasonable use case for this, so there should be some way to silence the warning at the point of use. Consider, for instance:

```
let physicalConstantA = /* some large number */
let physicalConstantB = physicalConstantA
```

Suppose `physicalConstantA` serves a different purpose and is used in different kinds of calculation than `physicalConstantB`; however, it is fixed to be the same value. With the warning above, I'm encouraged either to (A) refer to this value all the time as `physicalConstantA`, even where at the point of use this could be confusing because readers are accustomed to reasoning about a calculation in terms of `physicalConstantB`; or to (B) rewrite the large number twice, risking a second typo or the two values falling out of sync, which they should not.

In other words, the use case here would be analogous to `type` aliases, only in this case for values. I should be allowed to _alias_ immutable values, and so in the case where this is not being done accidentally, there should be a way to shut off this warning. Usually, parens are offered as a way to silence warnings like this.

If this produces a concern about how downstream diagnostics, we could work out some way of referring to such "value aliases" in diagnostics (for example, an imaginary error that the value is too amazing to be used might read: `'x' (aka 'y') is too amazing to be used`).

2. The wording of the diagnostic might be somewhat confusing. The user-facing language distinguishes "reference types" from "value types," and _semantically_, `let y = x` creates a second _copy_ of the same value. Now, of course, users familiar with compiler internals know that this "copy" is entirely notional and always elided when you've got immutable bindings of a constant value. However, calling these bindings "references" and telling the user that they've "referenced" the same value twice would be confusing to a user who's working with the mental model that Swift explicitly teaches them to use.

xwu commented:
> Because inactive clauses are not generated in SIL, we can not track usage here.
> 
> ```
> var x = 12 // we end up diagnosing this is never mutated
> #if false
>   x = 1
> #endif
> print(x)
> ```
>
> Future work may be done in SILGen to generate useful information about variable usage in inactive clauses if this is important. Or maybe some information can be inspected in the AST.
> Otherwise, this current diagnostic might encourage Swift users to write code with better usage enforcement across all builds.

Given the nature of the diagnostic in question, we should prefer false negatives (not showing the diagnostic when a variable isn't mutated) over false positives (showing the diagnostic incorrectly when a variable really is mutated).

Therefore--and I apologize if I misunderstand your text and this is already the behavior you've implemented--in the case where there's code that's conditionally not compiled, I think the diagnostic should just not be shown. This doesn't preclude the future work you mention where variable usage could be determined within these conditional compilation blocks, but until that time, I would fail open.

maustinstar commented:
@xwu I really like that idea for using parenthesis to explicitly note a copy of the reference. I will look into this.
***
For #if statements, you are correct in assuming that the diagnostic is shown if a variable is used within the conditional compilation blocks.

```swift
let a = 0 // diagnoses never used
#if false
print(a)
#endif 
```

The code for the inactive clause never goes through SILGen, so it's invisible to the SIL pass for diagnostics. So, I'm not sure how to find that a value is used in a case like this.

```swift
let a = 0 // ok
#if true
print(a)
#endif 
```

This code does pass through SILGen, so we can see this use in the diagnostic pass.

The only way I can imagine doing this is by changing SILGen (perhaps create some debug instruction to inform us that a value is captured in an inactive clause). Unfortunately, I'm not familiar with the SILGen process, but I will try my best to figure this out. Any ideas here?

xwu commented:
> The only way I can imagine doing this is by changing SILGen (perhaps create some debug instruction to inform us that a value is captured in an inactive clause). Unfortunately, I'm not familiar with the SILGen process, but I will try my best to figure this out. Any ideas here?

This does sound like it'd be a bit of a tough thing to tackle. To be clear, I'm not suggesting that it needs to be done as part of this PR even though it'd be a very nice improvement later. I'm only suggesting that, for now, we _don't_ diagnose unused variables in cases where there are conditional clauses (that is, let the presence of a conditional clause cause all in-scope variables to be considered "used" for now).

maustinstar commented:
Leaving an update so this thread doesn't go stale:

> The reason `_ = foo` produces no code is because SILGen peepholes it away. Perhaps we should just not do that, and allow dead code elimination to remove the discarded result after the diagnostics pass?

I started working on a change to SILGen to produce debug values for discarded assignment expressions. Otherwise, the use of the value is invisible by the time we get to this diagnostic pass. I am considering introducing this as an incremental change in a separate PR, since it seems other aspects of this project need further consideration and may go on for a long while.

In the meantime, I will convert this PR to a draft.

Thank you for the continuous feedback and support.

maustinstar committed:
eliminate SILGen peephole for discarded assignments

maustinstar committed:
Update tests

maustinstar committed:
Update Tests

maustinstar commented:
Update: was able to remove SILGen peephole for underscore assignments. This restores functionality for underscore assignments.

Tagging @varungandhi-apple, thanks for the help.

maustinstar committed:
Exempt if-config vars from diagnostics

maustinstar committed:
Create MarkDiscardedInst to track var usage

maustinstar committed:
Update tests

maustinstar commented:
@xwu, the regressions we discussed earlier in this thread have now been addressed. I would appreciate if you could run the test suite on CI.

maustinstar commented:
@CodaFi can you run this on CI?

maustinstar committed:
Remove DebugValueAddrInst from DiagnoseVarUsage

varungandhi-apple commented:
@swift-ci test

varungandhi-apple commented:
[Okay if other people who have studied this PR in more detail would like to overrule this but] Given the large scale of this change, I think there is a greater risk of things going wrong and that this may need to be reverted. Should that happen, the size of this PR creates a greater chance of merge conflicts, making reverting and relanding this work tricky.

If possible, I recommend breaking it up into at least 3 PRs, spaced out by 1-2 weeks:

PR 1. Add the new checker under a new frontend flag. Make sure any tests pass with both by default and with the new flag. (Alternately, this flag takes a boolean which is false by default). This is also a good opportunity to squash some of the WIP commits.
PR 2. Flip the default.
PR 3. Remove the old code and the flag.

maustinstar commented:
Okay, thanks @varungandhi-apple, I'll split this up and add the flag
